{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88c230a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "0257cdf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "daed8d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_table_a3026 = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/a3026_allele_table.csv',header=0,index_col=0,na_filter=False)\n",
    "allele_table_a3028 = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/a3028_allele_table.csv',header=0,index_col=0,na_filter=False)\n",
    "allele_table_a3030 = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/a3030_allele_table.csv',header=0,index_col=0,na_filter=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "76ee4706",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_table = pd.concat([allele_table_a3026,allele_table_a3028,allele_table_a3030],axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "483df507",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellBC</th>\n",
       "      <th>intBC</th>\n",
       "      <th>r1</th>\n",
       "      <th>allele</th>\n",
       "      <th>readCount</th>\n",
       "      <th>UMI</th>\n",
       "      <th>UMI_ratio</th>\n",
       "      <th>lineageGrp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5648</td>\n",
       "      <td>967</td>\n",
       "      <td>0.967000</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Btf3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>827</td>\n",
       "      <td>96</td>\n",
       "      <td>0.989691</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Calm2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3075</td>\n",
       "      <td>88</td>\n",
       "      <td>0.897959</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Hint1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>624</td>\n",
       "      <td>70</td>\n",
       "      <td>0.958904</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Slc25a3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>882</td>\n",
       "      <td>64</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496555</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496556</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Calr</td>\n",
       "      <td>81:2D</td>\n",
       "      <td>81:2D</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496557</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>93:1I</td>\n",
       "      <td>93:1I</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496558</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Rpl39</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496559</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_AAGTGAACAGGATCTT</td>\n",
       "      <td>Rpl39</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1535366 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Sample                         cellBC    intBC     r1 allele  \\\n",
       "0       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC     Tpt1   None   None   \n",
       "1       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC     Btf3   None   None   \n",
       "2       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC    Calm2   None   None   \n",
       "3       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC    Hint1   None   None   \n",
       "4       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC  Slc25a3   None   None   \n",
       "...              ...                            ...      ...    ...    ...   \n",
       "496555      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA     Tpt1   None   None   \n",
       "496556      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA     Calr  81:2D  81:2D   \n",
       "496557      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA     Tpt1  93:1I  93:1I   \n",
       "496558      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA    Rpl39   None   None   \n",
       "496559      T1_a3030      T1_a3030_AAGTGAACAGGATCTT    Rpl39   None   None   \n",
       "\n",
       "        readCount  UMI  UMI_ratio  lineageGrp  \n",
       "0            5648  967   0.967000           9  \n",
       "1             827   96   0.989691           9  \n",
       "2            3075   88   0.897959           9  \n",
       "3             624   70   0.958904           9  \n",
       "4             882   64   1.000000           9  \n",
       "...           ...  ...        ...         ...  \n",
       "496555         12    3   0.300000         129  \n",
       "496556        165    2   0.500000         129  \n",
       "496557         13    2   0.200000         129  \n",
       "496558          7    2   0.666667         129  \n",
       "496559         20    4   1.000000          12  \n",
       "\n",
       "[1535366 rows x 9 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "752fc7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = ['Acta1', 'Hsp90b1','Calr','Cfl1','Myl6','Rpl39','Calm2','Fth1','Hsp90ab1','Tpt1','Slc25a3','Lgals1','Btf3','Hint1','Calm1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "c1b0367f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellBC</th>\n",
       "      <th>intBC</th>\n",
       "      <th>r1</th>\n",
       "      <th>allele</th>\n",
       "      <th>readCount</th>\n",
       "      <th>UMI</th>\n",
       "      <th>UMI_ratio</th>\n",
       "      <th>lineageGrp</th>\n",
       "      <th>gene_r1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5648</td>\n",
       "      <td>967</td>\n",
       "      <td>0.967000</td>\n",
       "      <td>9</td>\n",
       "      <td>Tpt1_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Btf3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>827</td>\n",
       "      <td>96</td>\n",
       "      <td>0.989691</td>\n",
       "      <td>9</td>\n",
       "      <td>Btf3_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Calm2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3075</td>\n",
       "      <td>88</td>\n",
       "      <td>0.897959</td>\n",
       "      <td>9</td>\n",
       "      <td>Calm2_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Hint1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>624</td>\n",
       "      <td>70</td>\n",
       "      <td>0.958904</td>\n",
       "      <td>9</td>\n",
       "      <td>Hint1_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Slc25a3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>882</td>\n",
       "      <td>64</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9</td>\n",
       "      <td>Slc25a3_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496555</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>129</td>\n",
       "      <td>Tpt1_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496556</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Calr</td>\n",
       "      <td>81:2D</td>\n",
       "      <td>81:2D</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>129</td>\n",
       "      <td>Calr_81:2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496557</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>93:1I</td>\n",
       "      <td>93:1I</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>129</td>\n",
       "      <td>Tpt1_93:1I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496558</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Rpl39</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>129</td>\n",
       "      <td>Rpl39_None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496559</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_AAGTGAACAGGATCTT</td>\n",
       "      <td>Rpl39</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12</td>\n",
       "      <td>Rpl39_None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1535366 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Sample                         cellBC    intBC     r1 allele  \\\n",
       "0       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC     Tpt1   None   None   \n",
       "1       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC     Btf3   None   None   \n",
       "2       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC    Calm2   None   None   \n",
       "3       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC    Hint1   None   None   \n",
       "4       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC  Slc25a3   None   None   \n",
       "...              ...                            ...      ...    ...    ...   \n",
       "496555      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA     Tpt1   None   None   \n",
       "496556      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA     Calr  81:2D  81:2D   \n",
       "496557      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA     Tpt1  93:1I  93:1I   \n",
       "496558      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA    Rpl39   None   None   \n",
       "496559      T1_a3030      T1_a3030_AAGTGAACAGGATCTT    Rpl39   None   None   \n",
       "\n",
       "        readCount  UMI  UMI_ratio  lineageGrp       gene_r1  \n",
       "0            5648  967   0.967000           9     Tpt1_None  \n",
       "1             827   96   0.989691           9     Btf3_None  \n",
       "2            3075   88   0.897959           9    Calm2_None  \n",
       "3             624   70   0.958904           9    Hint1_None  \n",
       "4             882   64   1.000000           9  Slc25a3_None  \n",
       "...           ...  ...        ...         ...           ...  \n",
       "496555         12    3   0.300000         129     Tpt1_None  \n",
       "496556        165    2   0.500000         129    Calr_81:2D  \n",
       "496557         13    2   0.200000         129    Tpt1_93:1I  \n",
       "496558          7    2   0.666667         129    Rpl39_None  \n",
       "496559         20    4   1.000000          12    Rpl39_None  \n",
       "\n",
       "[1535366 rows x 10 columns]"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_table['gene_r1'] = allele_table['intBC'] + '_' + allele_table['r1']\n",
    "allele_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "bd37845d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>cellBC</th>\n",
       "      <th>intBC</th>\n",
       "      <th>r1</th>\n",
       "      <th>allele</th>\n",
       "      <th>readCount</th>\n",
       "      <th>UMI</th>\n",
       "      <th>UMI_ratio</th>\n",
       "      <th>lineageGrp</th>\n",
       "      <th>gene_r1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Rpl39</td>\n",
       "      <td>88:3D</td>\n",
       "      <td>88:3D</td>\n",
       "      <td>289</td>\n",
       "      <td>61</td>\n",
       "      <td>0.398693</td>\n",
       "      <td>9</td>\n",
       "      <td>Rpl39_88:3D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Rpl39</td>\n",
       "      <td>87:4D</td>\n",
       "      <td>87:4D</td>\n",
       "      <td>228</td>\n",
       "      <td>53</td>\n",
       "      <td>0.346405</td>\n",
       "      <td>9</td>\n",
       "      <td>Rpl39_87:4D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Myl6</td>\n",
       "      <td>84:9D</td>\n",
       "      <td>84:9D</td>\n",
       "      <td>315</td>\n",
       "      <td>38</td>\n",
       "      <td>0.481013</td>\n",
       "      <td>9</td>\n",
       "      <td>Myl6_84:9D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Myl6</td>\n",
       "      <td>90:2D</td>\n",
       "      <td>90:2D</td>\n",
       "      <td>271</td>\n",
       "      <td>28</td>\n",
       "      <td>0.354430</td>\n",
       "      <td>9</td>\n",
       "      <td>Myl6_90:2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T3_a3026_1_1</td>\n",
       "      <td>T3_a3026_1_1_CTACCCAAGGGCTTCC</td>\n",
       "      <td>Hsp90ab1</td>\n",
       "      <td>87:1D</td>\n",
       "      <td>87:1D</td>\n",
       "      <td>139</td>\n",
       "      <td>21</td>\n",
       "      <td>0.656250</td>\n",
       "      <td>9</td>\n",
       "      <td>Hsp90ab1_87:1D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496552</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>81:16D</td>\n",
       "      <td>81:16D</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>129</td>\n",
       "      <td>Tpt1_81:16D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496553</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Fth1</td>\n",
       "      <td>89:16D</td>\n",
       "      <td>89:16D</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>129</td>\n",
       "      <td>Fth1_89:16D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496554</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Fth1</td>\n",
       "      <td>90:1I</td>\n",
       "      <td>90:1I</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>129</td>\n",
       "      <td>Fth1_90:1I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496556</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Calr</td>\n",
       "      <td>81:2D</td>\n",
       "      <td>81:2D</td>\n",
       "      <td>165</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>129</td>\n",
       "      <td>Calr_81:2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496557</th>\n",
       "      <td>T1_a3030</td>\n",
       "      <td>T1_a3030_CAGTTAGTCAAACCCA</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>93:1I</td>\n",
       "      <td>93:1I</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>129</td>\n",
       "      <td>Tpt1_93:1I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>800695 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Sample                         cellBC     intBC      r1  allele  \\\n",
       "5       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC     Rpl39   88:3D   88:3D   \n",
       "6       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC     Rpl39   87:4D   87:4D   \n",
       "8       T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC      Myl6   84:9D   84:9D   \n",
       "10      T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC      Myl6   90:2D   90:2D   \n",
       "12      T3_a3026_1_1  T3_a3026_1_1_CTACCCAAGGGCTTCC  Hsp90ab1   87:1D   87:1D   \n",
       "...              ...                            ...       ...     ...     ...   \n",
       "496552      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA      Tpt1  81:16D  81:16D   \n",
       "496553      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA      Fth1  89:16D  89:16D   \n",
       "496554      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA      Fth1   90:1I   90:1I   \n",
       "496556      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA      Calr   81:2D   81:2D   \n",
       "496557      T1_a3030      T1_a3030_CAGTTAGTCAAACCCA      Tpt1   93:1I   93:1I   \n",
       "\n",
       "        readCount  UMI  UMI_ratio  lineageGrp         gene_r1  \n",
       "5             289   61   0.398693           9     Rpl39_88:3D  \n",
       "6             228   53   0.346405           9     Rpl39_87:4D  \n",
       "8             315   38   0.481013           9      Myl6_84:9D  \n",
       "10            271   28   0.354430           9      Myl6_90:2D  \n",
       "12            139   21   0.656250           9  Hsp90ab1_87:1D  \n",
       "...           ...  ...        ...         ...             ...  \n",
       "496552         18    4   0.400000         129     Tpt1_81:16D  \n",
       "496553         18    3   0.500000         129     Fth1_89:16D  \n",
       "496554         13    3   0.500000         129      Fth1_90:1I  \n",
       "496556        165    2   0.500000         129      Calr_81:2D  \n",
       "496557         13    2   0.200000         129      Tpt1_93:1I  \n",
       "\n",
       "[800695 rows x 10 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_table_mut = allele_table[allele_table['r1']!='None']\n",
    "allele_table_mut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "c66f72b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5292"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(allele_table_mut['gene_r1'][allele_table_mut['intBC']!='Acta1']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "b1e6cf70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{5, 14, 47, 62}"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(allele_table_mut['lineageGrp'][allele_table_mut['intBC']=='Acta1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34b98775",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ca0eb2b5",
   "metadata": {},
   "source": [
    "# T1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ce7d5744",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = ['T1_a3026','T1_a3028','T1_a3030']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d0266966",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acta1\n",
      "Hsp90b1\n",
      "Calr\n",
      "Cfl1\n",
      "Myl6\n",
      "Rpl39\n",
      "Calm2\n",
      "Fth1\n",
      "Hsp90ab1\n",
      "Tpt1\n",
      "Slc25a3\n",
      "Lgals1\n",
      "Btf3\n",
      "Hint1\n",
      "Calm1\n"
     ]
    }
   ],
   "source": [
    "allele_table_rep = allele_table[allele_table['Sample'].isin(samples)]\n",
    "cell_mutation_matrix = pd.DataFrame(index=list(set(allele_table_rep['cellBC'])))\n",
    "for gene in gene_names:\n",
    "    print(gene)\n",
    "    allele_table_gene = allele_table_rep[allele_table_rep['intBC']==gene]\n",
    "    allele_table_gene = allele_table_gene[allele_table_gene['r1']!='None']\n",
    "    df = allele_table_gene.pivot(index='cellBC', columns='r1', values='UMI')\n",
    "    df= df.fillna(0)\n",
    "    df[df > 0] = 1\n",
    "    df.columns = [gene + '_' + indel for indel in df.columns]\n",
    "    cell_mutation_matrix = cell_mutation_matrix.join(df, how='outer')\n",
    "    cell_mutation_matrix = cell_mutation_matrix.fillna(0)\n",
    "\n",
    "cell_lineage = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/cell_lineage.T1_T2_T3a.csv',header=0)\n",
    "cell_lineage.index = cell_lineage.cellBC\n",
    "cell_lineage = cell_lineage.loc[allele_table_rep.cellBC.unique(),:]\n",
    "\n",
    "for lineageGrp in cell_lineage['lineageGrp'].unique():\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix.loc[cell_lineage[cell_lineage['lineageGrp'].isin([lineageGrp])]['cellBC'],:]           \n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.astype(int)\n",
    "    if cell_mutation_matrix_lineageGrp.shape[0] < 100:\n",
    "        continue\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.loc[:, (cell_mutation_matrix_lineageGrp != 0).any(axis=0)]\n",
    "    if cell_mutation_matrix_lineageGrp.shape[1] == 0:\n",
    "        continue\n",
    "    cell_mutation_matrix_lineageGrp.loc['synthetic'] = [ 0 for i in range(0,cell_mutation_matrix_lineageGrp.shape[1])]\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp.to_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T1/'+'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.csv')\n",
    "\n",
    "    cell_mutation_sequence = cell_mutation_matrix_lineageGrp.apply(lambda row: ''.join(row.astype(str)), axis=1)\n",
    "    output_file = '/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T1/'+ 'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.phy'\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.write(f\"{cell_mutation_matrix_lineageGrp.shape[0]} \\t{cell_mutation_matrix_lineageGrp.shape[1]}\\n\") \n",
    "        for index, line in cell_mutation_sequence.items():\n",
    "            file.write(f\"{index}\\t{line}\\n\")  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d70a2a2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "04b95e37",
   "metadata": {},
   "source": [
    "# T2T3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d75a923f",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = ['T2_a3026_1','T2_a3026_2','T3_a3026_1_1','T3_a3026_2_1','T2_a3028_1','T2_a3028_2','T3_a3028_1_1','T3_a3028_2_1','T2_a3030_1','T2_a3030_2','T3_a3030_1_1','T3_a3030_2_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feea9c2d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a929c07b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acta1\n",
      "Hsp90b1\n",
      "Calr\n",
      "Cfl1\n",
      "Myl6\n",
      "Rpl39\n",
      "Calm2\n",
      "Fth1\n",
      "Hsp90ab1\n",
      "Tpt1\n",
      "Slc25a3\n",
      "Lgals1\n",
      "Btf3\n",
      "Hint1\n",
      "Calm1\n"
     ]
    }
   ],
   "source": [
    "allele_table_rep = allele_table[allele_table['Sample'].isin(samples)]\n",
    "cell_mutation_matrix = pd.DataFrame(index=list(set(allele_table_rep['cellBC'])))\n",
    "for gene in gene_names:\n",
    "    print(gene)\n",
    "    allele_table_gene = allele_table_rep[allele_table_rep['intBC']==gene]\n",
    "    allele_table_gene = allele_table_gene[allele_table_gene['r1']!='None']\n",
    "    df = allele_table_gene.pivot(index='cellBC', columns='r1', values='UMI')\n",
    "    df= df.fillna(0)\n",
    "    df[df > 0] = 1\n",
    "    df.columns = [gene + '_' + indel for indel in df.columns]\n",
    "    cell_mutation_matrix = cell_mutation_matrix.join(df, how='outer')\n",
    "    cell_mutation_matrix = cell_mutation_matrix.fillna(0)\n",
    "\n",
    "cell_lineage = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/cell_lineage.T1_T2_T3a.csv',header=0)\n",
    "cell_lineage.index = cell_lineage.cellBC\n",
    "cell_lineage = cell_lineage.loc[allele_table_rep.cellBC.unique(),:]\n",
    "\n",
    "for lineageGrp in cell_lineage['lineageGrp'].unique():\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix.loc[cell_lineage[cell_lineage['lineageGrp'].isin([lineageGrp])]['cellBC'],:]           \n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.astype(int)\n",
    "    if cell_mutation_matrix_lineageGrp.shape[0] < 10:\n",
    "        continue\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.loc[:, (cell_mutation_matrix_lineageGrp != 0).any(axis=0)]\n",
    "    if cell_mutation_matrix_lineageGrp.shape[1] == 0:\n",
    "        continue\n",
    "    cell_mutation_matrix_lineageGrp.loc['synthetic'] = [ 0 for i in range(0,cell_mutation_matrix_lineageGrp.shape[1])]\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp.to_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T2_and_T3/'+'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.csv')\n",
    "\n",
    "    cell_mutation_sequence = cell_mutation_matrix_lineageGrp.apply(lambda row: ''.join(row.astype(str)), axis=1)\n",
    "    output_file = '/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T2_and_T3/'+ 'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.phy'\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.write(f\"{cell_mutation_matrix_lineageGrp.shape[0]} \\t{cell_mutation_matrix_lineageGrp.shape[1]}\\n\") \n",
    "        for index, line in cell_mutation_sequence.items():\n",
    "            file.write(f\"{index}\\t{line}\\n\")  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b761137",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e8b97304",
   "metadata": {},
   "source": [
    "# T1T2T3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e5cb1f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = ['T1_a3026','T1_a3028','T1_a3030','T2_a3026_1','T2_a3026_2','T3_a3026_1_1','T3_a3026_2_1','T2_a3028_1','T2_a3028_2','T3_a3028_1_1','T3_a3028_2_1','T2_a3030_1','T2_a3030_2','T3_a3030_1_1','T3_a3030_2_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e2e032d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acta1\n",
      "Hsp90b1\n",
      "Calr\n",
      "Cfl1\n",
      "Myl6\n",
      "Rpl39\n",
      "Calm2\n",
      "Fth1\n",
      "Hsp90ab1\n",
      "Tpt1\n",
      "Slc25a3\n",
      "Lgals1\n",
      "Btf3\n",
      "Hint1\n",
      "Calm1\n"
     ]
    }
   ],
   "source": [
    "allele_table_rep = allele_table[allele_table['Sample'].isin(samples)]\n",
    "cell_mutation_matrix = pd.DataFrame(index=list(set(allele_table_rep['cellBC'])))\n",
    "for gene in gene_names:\n",
    "    print(gene)\n",
    "    allele_table_gene = allele_table_rep[allele_table_rep['intBC']==gene]\n",
    "    allele_table_gene = allele_table_gene[allele_table_gene['r1']!='None']\n",
    "    df = allele_table_gene.pivot(index='cellBC', columns='r1', values='UMI')\n",
    "    df= df.fillna(0)\n",
    "    df[df > 0] = 1\n",
    "    df.columns = [gene + '_' + indel for indel in df.columns]\n",
    "    cell_mutation_matrix = cell_mutation_matrix.join(df, how='outer')\n",
    "    cell_mutation_matrix = cell_mutation_matrix.fillna(0)\n",
    "\n",
    "cell_lineage = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/cell_lineage.T1_T2_T3a.csv',header=0)\n",
    "cell_lineage.index = cell_lineage.cellBC\n",
    "cell_lineage = cell_lineage.loc[allele_table_rep.cellBC.unique(),:]\n",
    "\n",
    "for lineageGrp in cell_lineage['lineageGrp'].unique():\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix.loc[cell_lineage[cell_lineage['lineageGrp'].isin([lineageGrp])]['cellBC'],:]           \n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.astype(int)\n",
    "    if cell_mutation_matrix_lineageGrp.shape[0] < 10:\n",
    "        continue\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.loc[:, (cell_mutation_matrix_lineageGrp != 0).any(axis=0)]\n",
    "    if cell_mutation_matrix_lineageGrp.shape[1] == 0:\n",
    "        continue\n",
    "    cell_mutation_matrix_lineageGrp.loc['synthetic'] = [ 0 for i in range(0,cell_mutation_matrix_lineageGrp.shape[1])]\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp.to_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T1T2T3/'+'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.csv')\n",
    "\n",
    "    cell_mutation_sequence = cell_mutation_matrix_lineageGrp.apply(lambda row: ''.join(row.astype(str)), axis=1)\n",
    "    output_file = '/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T1T2T3/'+ 'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.phy'\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.write(f\"{cell_mutation_matrix_lineageGrp.shape[0]} \\t{cell_mutation_matrix_lineageGrp.shape[1]}\\n\") \n",
    "        for index, line in cell_mutation_sequence.items():\n",
    "            file.write(f\"{index}\\t{line}\\n\")  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4014a77e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'T1_a3026',\n",
       " 'T1_a3028',\n",
       " 'T1_a3030',\n",
       " 'T2_a3026_1',\n",
       " 'T2_a3026_2',\n",
       " 'T2_a3028_1',\n",
       " 'T2_a3028_2',\n",
       " 'T2_a3030_1',\n",
       " 'T2_a3030_2',\n",
       " 'T3_a3026_1_1',\n",
       " 'T3_a3026_2_1',\n",
       " 'T3_a3028_1_1',\n",
       " 'T3_a3028_2_1',\n",
       " 'T3_a3030_1_1',\n",
       " 'T3_a3030_2_1'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(allele_table_rep['Sample'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f3f9627",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "480d1300",
   "metadata": {},
   "outputs": [],
   "source": [
    "clones = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/T1T2T3/clone.mut.list',header=None)\n",
    "cellbc_all = []\n",
    "for file_phy in clones[0]:\n",
    "    phy = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/T1T2T3/'+file_phy,sep='\\t')\n",
    "    cellbc_all = cellbc_all + phy.iloc[:,0].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fcd39ac5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70037"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cellbc_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5fe3ab54",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_351684/2648883166.py:1: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metadata = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/scRNA/cellranger/metadata.T1_T2_T3a.harmony.final.refine.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>RNA_snn_res.0.2</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>pANN_0.25_0.12_592</th>\n",
       "      <th>Doublet_or_Singlet</th>\n",
       "      <th>pANN_0.25_0.21_600</th>\n",
       "      <th>...</th>\n",
       "      <th>pANN_0.25_0.29_759</th>\n",
       "      <th>pANN_0.25_0.28_612</th>\n",
       "      <th>pANN_0.25_0.14_771</th>\n",
       "      <th>pANN_0.25_0.07_798</th>\n",
       "      <th>RNA_snn_res.0.3</th>\n",
       "      <th>harmony_clusters</th>\n",
       "      <th>seurat_clusters_rename</th>\n",
       "      <th>umapharmony_1</th>\n",
       "      <th>umapharmony_2</th>\n",
       "      <th>cellName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T1_T1_a3026_AAACCCAAGACTAGAT-1</td>\n",
       "      <td>T1_a3026</td>\n",
       "      <td>23101</td>\n",
       "      <td>4348</td>\n",
       "      <td>5.653435</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.314012</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3.960447</td>\n",
       "      <td>5.419786</td>\n",
       "      <td>T1_a3026_AAACCCAAGACTAGAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T1_T1_a3026_AAACCCACATGAGATA-1</td>\n",
       "      <td>T1_a3026</td>\n",
       "      <td>13159</td>\n",
       "      <td>3546</td>\n",
       "      <td>2.029030</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.356351</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-4.321606</td>\n",
       "      <td>-2.761532</td>\n",
       "      <td>T1_a3026_AAACCCACATGAGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T1_T1_a3026_AAACCCAGTCGCTGCA-1</td>\n",
       "      <td>T1_a3026</td>\n",
       "      <td>15441</td>\n",
       "      <td>3362</td>\n",
       "      <td>5.984068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.256048</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.283126</td>\n",
       "      <td>-2.055104</td>\n",
       "      <td>T1_a3026_AAACCCAGTCGCTGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T1_T1_a3026_AAACCCAGTCTCGCGA-1</td>\n",
       "      <td>T1_a3026</td>\n",
       "      <td>87500</td>\n",
       "      <td>7440</td>\n",
       "      <td>4.874286</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.452117</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.347915</td>\n",
       "      <td>-1.723554</td>\n",
       "      <td>T1_a3026_AAACCCAGTCTCGCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T1_T1_a3026_AAACCCATCACGGGAA-1</td>\n",
       "      <td>T1_a3026</td>\n",
       "      <td>22915</td>\n",
       "      <td>3939</td>\n",
       "      <td>4.233035</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.880527</td>\n",
       "      <td>-2.335177</td>\n",
       "      <td>T1_a3026_AAACCCATCACGGGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91406</th>\n",
       "      <td>T3_T3_a3030_TTTGTTGCATGGACAG-1</td>\n",
       "      <td>T3_a3030_1_1</td>\n",
       "      <td>21144</td>\n",
       "      <td>4395</td>\n",
       "      <td>7.609724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.269478</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>-1.273801</td>\n",
       "      <td>-2.477038</td>\n",
       "      <td>T3_a3030_1_1_TTTGTTGCATGGACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91407</th>\n",
       "      <td>T3_T3_a3030_TTTGTTGGTAAGATCA-1</td>\n",
       "      <td>T3_a3030_2_1</td>\n",
       "      <td>15115</td>\n",
       "      <td>4228</td>\n",
       "      <td>5.339067</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.369386</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.885944</td>\n",
       "      <td>-1.561220</td>\n",
       "      <td>T3_a3030_2_1_TTTGTTGGTAAGATCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91408</th>\n",
       "      <td>T3_T3_a3030_TTTGTTGGTGCGGTAA-1</td>\n",
       "      <td>T3_a3030_1_1</td>\n",
       "      <td>12252</td>\n",
       "      <td>2836</td>\n",
       "      <td>5.811296</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.109991</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-5.376790</td>\n",
       "      <td>-0.871353</td>\n",
       "      <td>T3_a3030_1_1_TTTGTTGGTGCGGTAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91409</th>\n",
       "      <td>T3_T3_a3030_TTTGTTGTCCGCAAAT-1</td>\n",
       "      <td>T3_a3030_2_1</td>\n",
       "      <td>17460</td>\n",
       "      <td>4159</td>\n",
       "      <td>4.444444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.206233</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.321478</td>\n",
       "      <td>-1.859076</td>\n",
       "      <td>T3_a3030_2_1_TTTGTTGTCCGCAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91410</th>\n",
       "      <td>T3_T3_a3030_TTTGTTGTCGGTATGT-1</td>\n",
       "      <td>T3_a3030_2_1</td>\n",
       "      <td>33977</td>\n",
       "      <td>6577</td>\n",
       "      <td>3.134473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Singlet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.349221</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3.145834</td>\n",
       "      <td>4.822145</td>\n",
       "      <td>T3_a3030_2_1_TTTGTTGTCGGTATGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91411 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Unnamed: 0    orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "0      T1_T1_a3026_AAACCCAAGACTAGAT-1      T1_a3026       23101          4348   \n",
       "1      T1_T1_a3026_AAACCCACATGAGATA-1      T1_a3026       13159          3546   \n",
       "2      T1_T1_a3026_AAACCCAGTCGCTGCA-1      T1_a3026       15441          3362   \n",
       "3      T1_T1_a3026_AAACCCAGTCTCGCGA-1      T1_a3026       87500          7440   \n",
       "4      T1_T1_a3026_AAACCCATCACGGGAA-1      T1_a3026       22915          3939   \n",
       "...                               ...           ...         ...           ...   \n",
       "91406  T3_T3_a3030_TTTGTTGCATGGACAG-1  T3_a3030_1_1       21144          4395   \n",
       "91407  T3_T3_a3030_TTTGTTGGTAAGATCA-1  T3_a3030_2_1       15115          4228   \n",
       "91408  T3_T3_a3030_TTTGTTGGTGCGGTAA-1  T3_a3030_1_1       12252          2836   \n",
       "91409  T3_T3_a3030_TTTGTTGTCCGCAAAT-1  T3_a3030_2_1       17460          4159   \n",
       "91410  T3_T3_a3030_TTTGTTGTCGGTATGT-1  T3_a3030_2_1       33977          6577   \n",
       "\n",
       "       percent.mt  RNA_snn_res.0.2  seurat_clusters  pANN_0.25_0.12_592  \\\n",
       "0        5.653435              1.0                1            0.314012   \n",
       "1        2.029030              0.0                2            0.356351   \n",
       "2        5.984068              0.0                2            0.256048   \n",
       "3        4.874286              0.0                2            0.452117   \n",
       "4        4.233035              0.0                2            0.250000   \n",
       "...           ...              ...              ...                 ...   \n",
       "91406    7.609724              NaN                0                 NaN   \n",
       "91407    5.339067              NaN                0                 NaN   \n",
       "91408    5.811296              NaN                2                 NaN   \n",
       "91409    4.444444              NaN                0                 NaN   \n",
       "91410    3.134473              NaN                1                 NaN   \n",
       "\n",
       "      Doublet_or_Singlet  pANN_0.25_0.21_600  ...  pANN_0.25_0.29_759  \\\n",
       "0                Singlet                 NaN  ...                 NaN   \n",
       "1                Singlet                 NaN  ...                 NaN   \n",
       "2                Singlet                 NaN  ...                 NaN   \n",
       "3                Singlet                 NaN  ...                 NaN   \n",
       "4                Singlet                 NaN  ...                 NaN   \n",
       "...                  ...                 ...  ...                 ...   \n",
       "91406            Singlet                 NaN  ...                 NaN   \n",
       "91407            Singlet                 NaN  ...                 NaN   \n",
       "91408            Singlet                 NaN  ...                 NaN   \n",
       "91409            Singlet                 NaN  ...                 NaN   \n",
       "91410            Singlet                 NaN  ...                 NaN   \n",
       "\n",
       "      pANN_0.25_0.28_612 pANN_0.25_0.14_771  pANN_0.25_0.07_798  \\\n",
       "0                    NaN                NaN                 NaN   \n",
       "1                    NaN                NaN                 NaN   \n",
       "2                    NaN                NaN                 NaN   \n",
       "3                    NaN                NaN                 NaN   \n",
       "4                    NaN                NaN                 NaN   \n",
       "...                  ...                ...                 ...   \n",
       "91406                NaN                NaN            0.269478   \n",
       "91407                NaN                NaN            0.369386   \n",
       "91408                NaN                NaN            0.109991   \n",
       "91409                NaN                NaN            0.206233   \n",
       "91410                NaN                NaN            0.349221   \n",
       "\n",
       "       RNA_snn_res.0.3  harmony_clusters  seurat_clusters_rename  \\\n",
       "0                    1                 1                       2   \n",
       "1                    0                 2                       1   \n",
       "2                    0                 2                       1   \n",
       "3                    0                 2                       1   \n",
       "4                    0                 2                       1   \n",
       "...                ...               ...                     ...   \n",
       "91406                3                 0                       3   \n",
       "91407                2                 0                       3   \n",
       "91408                4                 2                       1   \n",
       "91409                2                 0                       3   \n",
       "91410                1                 1                       2   \n",
       "\n",
       "       umapharmony_1  umapharmony_2                       cellName  \n",
       "0           3.960447       5.419786      T1_a3026_AAACCCAAGACTAGAT  \n",
       "1          -4.321606      -2.761532      T1_a3026_AAACCCACATGAGATA  \n",
       "2          -2.283126      -2.055104      T1_a3026_AAACCCAGTCGCTGCA  \n",
       "3          -2.347915      -1.723554      T1_a3026_AAACCCAGTCTCGCGA  \n",
       "4          -2.880527      -2.335177      T1_a3026_AAACCCATCACGGGAA  \n",
       "...              ...            ...                            ...  \n",
       "91406      -1.273801      -2.477038  T3_a3030_1_1_TTTGTTGCATGGACAG  \n",
       "91407       0.885944      -1.561220  T3_a3030_2_1_TTTGTTGGTAAGATCA  \n",
       "91408      -5.376790      -0.871353  T3_a3030_1_1_TTTGTTGGTGCGGTAA  \n",
       "91409       0.321478      -1.859076  T3_a3030_2_1_TTTGTTGTCCGCAAAT  \n",
       "91410       3.145834       4.822145  T3_a3030_2_1_TTTGTTGTCGGTATGT  \n",
       "\n",
       "[91411 rows x 26 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/scRNA/cellranger/metadata.T1_T2_T3a.harmony.final.refine.csv')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "34a5c6b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69625"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(metadata['cellName'][metadata['cellName'].isin(cellbc_all)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e9d6bd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e4bce690",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acta1\n",
      "Hsp90b1\n",
      "Calr\n",
      "Cfl1\n",
      "Myl6\n",
      "Rpl39\n",
      "Calm2\n",
      "Fth1\n",
      "Hsp90ab1\n",
      "Tpt1\n",
      "Slc25a3\n",
      "Lgals1\n",
      "Btf3\n",
      "Hint1\n",
      "Calm1\n"
     ]
    }
   ],
   "source": [
    "allele_table_rep = allele_table[allele_table['Sample'].isin(samples)]\n",
    "cell_mutation_matrix = pd.DataFrame(index=list(set(allele_table_rep['cellBC'])))\n",
    "for gene in gene_names:\n",
    "    print(gene)\n",
    "    allele_table_gene = allele_table_rep[allele_table_rep['intBC']==gene]\n",
    "    allele_table_gene = allele_table_gene[allele_table_gene['r1']!='None']\n",
    "    df = allele_table_gene.pivot(index='cellBC', columns='r1', values='UMI')\n",
    "    df= df.fillna(0)\n",
    "    df[df > 0] = 1\n",
    "    df.columns = [gene + '_' + indel for indel in df.columns]\n",
    "    cell_mutation_matrix = cell_mutation_matrix.join(df, how='outer')\n",
    "    cell_mutation_matrix = cell_mutation_matrix.fillna(0)\n",
    "\n",
    "cell_lineage = pd.read_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/cell_lineage.T1_T2_T3a.csv',header=0)\n",
    "cell_lineage.index = cell_lineage.cellBC\n",
    "cell_lineage = cell_lineage.loc[allele_table_rep.cellBC.unique(),:]\n",
    "\n",
    "for lineageGrp in cell_lineage['lineageGrp'].unique():\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix.loc[cell_lineage[cell_lineage['lineageGrp'].isin([lineageGrp])]['cellBC'],:]           \n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.astype(int)\n",
    "    if cell_mutation_matrix_lineageGrp.shape[0] < 10:\n",
    "        continue\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp = cell_mutation_matrix_lineageGrp.loc[(cell_mutation_matrix_lineageGrp != 0).any(axis=1), (cell_mutation_matrix_lineageGrp != 0).any(axis=0)]\n",
    "    if cell_mutation_matrix_lineageGrp.shape[1] == 0:\n",
    "        continue\n",
    "    cell_mutation_matrix_lineageGrp.loc['synthetic'] = [ 0 for i in range(0,cell_mutation_matrix_lineageGrp.shape[1])]\n",
    "\n",
    "    cell_mutation_matrix_lineageGrp.to_csv('/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T1T2T3_filtered/'+'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.csv')\n",
    "\n",
    "    cell_mutation_sequence = cell_mutation_matrix_lineageGrp.apply(lambda row: ''.join(row.astype(str)), axis=1)\n",
    "    output_file = '/syn1/liangzhen/jinhua_jilab_project/result/DNA_Amplicon/trees/merge_all_experiments/'+'T1T2T3_filtered/'+ 'cell_mutation_matrix-C'+str(lineageGrp)+'-'+str(cell_mutation_matrix_lineageGrp.shape[0])+'.phy'\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.write(f\"{cell_mutation_matrix_lineageGrp.shape[0]} \\t{cell_mutation_matrix_lineageGrp.shape[1]}\\n\") \n",
    "        for index, line in cell_mutation_sequence.items():\n",
    "            file.write(f\"{index}\\t{line}\\n\")  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7a19f01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6576b193",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cassiopeia2",
   "language": "python",
   "name": "cassiopeia2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
